### DISCUSSION TITLE
Need to update primary categories and remap old contents with new primary categories

### DISCUSSION THREAD

**Discussion by paritshivani** [Sunbird-Ed/Community#722] [Sunbird ED configuration discussion]
Hi team,
Hope you are doing well!
**We have a requirement from client where we want to remove following Primary Categories**
1. eTextbook
2. Explanation Content
3. Learning Resource
4. Practice Question Set
5. Teacher Resource
6. Exam Question

These are configured in upload content also.
**We want add new Primary categories as following**
1. Good Practices
2. Manuals/SOPs
3. Reports

**- Is it possible to change this ? If yes, How can we update these categories.?Any configurations should be updated? 
- Which steps we need to follow to update the primary Categories in portal? 
- What are the impacts of this change? How can we update old contents with old primary categories to new primary categories?
- Do we need to update anything on frontend or in editor or player according to above change? What will be code changes we need to do**

We appreciate any insights. Thank you!

@vinukumar-vs ,@rajeevsathish, @AmiableAnil , @Krishnaj20, @chitranshu-keshav, @SanthoshVasabhaktula, @gandham-santhosh , @coolbung @komalm @manishakapadnis

**Comment by paritshivani** [comment]
Hello falks!!!
Any update on this? 
We are waiting for you valuable insights.

**Reply by vinukumar-vs** [reply]
@shubhambansaltarento @ashokreddy1208 Can you share the form configuration CURL to update these properties.

@mohitga @anisha147

**Reply by vinukumar-vs** [reply]
@paritshivani All these categories are showing based on 'contentPrimaryCategories' added for the specific channel. 
After creating the creating the primary categories, append the new primary categories to the channel.
<img width="1489" alt="image" src="https://github.com/Sunbird-Ed/Community/assets/4496096/f6165d55-2b5f-483d-baa1-f4b74c795da1">

> curl --location --request PATCH '{{host}}/api/channel/v1/update/01269878797503692810' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer {{kong_api_key}}' \
--data '{
    "request": {
        "channel": {
            "contentPrimaryCategories": [
                "eTextbook",
                "Explanation Content",
                "Learning Resource",
                "Practice Question Set",
                "Teacher Resource",
                "Exam Question",
                "Testing"
            ],
            "publisher": [
                {
                    "value": "01269878797503692810"
                }
            ]
        }
    }
}'

I would suggest for any content related queries to rise the discussion thread in KnowlgBB discussions for more details on migration.

**Comment by ashokreddy1208** [comment]
@paritshivani Following are the steps to update its Primary categories and defination.

**Step 1:-** Create primary category using 'api/object/category/v1/create' api, below is the sample curl.
`curl --location '{{host}}/api/object/category/v1/create' \
--header 'Content-Type: application/json' \
--header 'Authorization:  {{APIKey}}' \
--data '{
    "request": {
        "objectCategory": {
            "name": "Testing",
            "description": "Testing"
        }
    }
}'`

**Step 2:-** Create Object category definition using 'api/object/category/definition/v1/create' api, below is the sample curl.
`curl --location '{{host}}/api/object/category/definition/v1/create' \
--header 'Content-Type: application/json' \
--header 'Authorization: {{APIKey}}' \
--data '{
    "request": {
        "objectCategoryDefinition": {
            "categoryId": "obj-cat:testing",
            "targetObjectType": "Content",
            "objectMetadata": {},
            "forms": {}
        }
    }
}'`
Note:- update objectMetadata and forms as per your requirement.

**Step3:-** Update primaryCategories using 'channel/v1/update/' api, below is the sample curl 
`curl --location --request PATCH '{{host}}/api/channel/v1/update/{{ORGID}}' \
--header 'Content-Type: application/json' \
--header 'Authorization: {{APIKey}}' \
--data '{
    "request": {
        "channel": {
            "contentPrimaryCategories": [
                "eTextbook",
                "Explanation Content",
                "Learning Resource",
                "Practice Question Set",
                "Teacher Resource",
                "Exam Question",
                "Testing"
            ]
        }
    }
}'`

You can remove existing 'PrimaryCategories' from channel update api and add new 'PrimaryCategories'  values.

cc @vinukumar-vs @mohitga @amitpriyadarshi @anisha147

**Reply by nidhikachhawa24** [reply]
@ashokreddy1208 
I followed the above steps, but step 3 is not working.

It's throwing error: 
"Please provide valid : [contentPrimaryCategories]"

This is body:
`{
    "request": {
        "channel": {
            "contentPrimaryCategories": [
                "Course Assessment",
                "eTextbook",
                "Explanation Content",
                "Learning Resource",
                "Practice Question Set",
                "Teacher Resource",
                "Exam Question",
                "Lesson Plan Unit"
            ]
        }
    }
}`

cc: @paritshivani

**Reply by nidhikachhawa24** [reply]
@ashokreddy1208 

this is response of my channel read: 
`{
    "id": "api.channel.read",
    "ver": "1.0",
    "ts": "2024-07-02T04:58:20.569Z",
    "params": {
        "resmsgid": "b3f6d090-382f-11ef-8660-2dbf170944ef",
        "msgid": "b3e9d840-382f-11ef-8b08-a5c253c2032a",
        "status": "successful",
        "err": null,
        "errmsg": null
    },
    "responseCode": "OK",
    "result": {
        "channel": {
            "identifier": "0140655149980794883",
            "lastStatusChangedOn": "2024-05-29T11:08:08.820+0000",
            "code": "0140655149980794883",
            "frameworks": [
                {
                    "name": "TestoneFramework",
                    "relation": "hasSequenceMember",
                    "identifier": "TestoneFramework",
                    "description": "TOFramework",
                    "objectType": "Framework",
                    "status": "Live",
                    "type": "K-12"
                }
            ],
            "consumerId": "52aa1fe8-3cb9-4f40-83a5-c3718c9edd8e",
            "channel": "in.ekstep",
            "description": "testing organisation for dev",
            "assetAdditionalCategories": [],
            "languageCode": [],
            "createdOn": "2024-05-29T11:08:08.820+0000",
            "objectType": "Channel",
            "versionKey": "1716980888820",
            "collectionPrimaryCategories": [
                "Content Playlist",
                "Course",
                "Digital Textbook",
                "Question paper"
            ],
            "contentPrimaryCategories": [
                "Course Assessment",
                "eTextbook",
                "Explanation Content",
                "Learning Resource",
                "Practice Question Set",
                "Teacher Resource",
                "Exam Question"
            ],
            "appId": "dev.nulp.portal",
            "name": "DevOneTestOrg",
            "primaryCategories": [
                {
                    "identifier": "obj-cat:asset_asset_all",
                    "name": "Asset",
                    "targetObjectType": "Asset"
                },
                {
                    "identifier": "obj-cat:certasset_asset_all",
                    "name": "CertAsset",
                    "targetObjectType": "Asset"
                },
                {
                    "identifier": "obj-cat:certificate-template_asset_all",
                    "name": "Certificate Template",
                    "targetObjectType": "Asset"
                },
                {
                    "identifier": "obj-cat:content-playlist_collection_all",
                    "name": "Content Playlist",
                    "targetObjectType": "Collection"
                },
                {
                    "identifier": "obj-cat:content-playlist_content_all",
                    "name": "Content Playlist",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:course_collection_all",
                    "name": "Course",
                    "targetObjectType": "Collection"
                },
                {
                    "identifier": "obj-cat:course_content_all",
                    "name": "Course",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:course-assessment_content_all",
                    "name": "Course Assessment",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:course-unit_content_all",
                    "name": "Course Unit",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:course-unit_collection_all",
                    "name": "Course Unit",
                    "targetObjectType": "Collection"
                },
                {
                    "identifier": "obj-cat:digital-textbook_content_all",
                    "name": "Digital Textbook",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:digital-textbook_collection_all",
                    "name": "Digital Textbook",
                    "targetObjectType": "Collection"
                },
                {
                    "identifier": "obj-cat:etextbook_content_all",
                    "name": "eTextbook",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:explanation-content_content_all",
                    "name": "Explanation Content",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:explanation-content_collection_all",
                    "name": "Explanation Content",
                    "targetObjectType": "Collection"
                },
                {
                    "identifier": "obj-cat:ftb-question_question_all",
                    "name": "FTB Question",
                    "targetObjectType": "Question"
                },
                {
                    "identifier": "obj-cat:goals_collection_all",
                    "name": "Goals",
                    "targetObjectType": "Collection"
                },
                {
                    "identifier": "obj-cat:good-practices_content_all",
                    "name": "Good Practices",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:learning-resource_content_all",
                    "name": "Learning Resource",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:lesson-plan-unit_content_all",
                    "name": "Lesson Plan Unit",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:lesson-plan-unit_collection_all",
                    "name": "Lesson Plan Unit",
                    "targetObjectType": "Collection"
                },
                {
                    "identifier": "obj-cat:manualssops_content_all",
                    "name": "Manuals/SOPs",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:multiple-choice-question_question_all",
                    "name": "Multiple Choice Question",
                    "targetObjectType": "Question"
                },
                {
                    "identifier": "obj-cat:playlist_collection_all",
                    "name": "Playlist",
                    "targetObjectType": "Collection"
                },
                {
                    "identifier": "obj-cat:plugin_content_all",
                    "name": "Plugin",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:practice-question-set_content_all",
                    "name": "Practice Question Set",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:reports_content_all",
                    "name": "Reports",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:subjective-question_question_all",
                    "name": "Subjective Question",
                    "targetObjectType": "Question"
                },
                {
                    "identifier": "obj-cat:teacher-resource_content_all",
                    "name": "Teacher Resource",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:template_content_all",
                    "name": "Template",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:textbook-unit_content_all",
                    "name": "Textbook Unit",
                    "targetObjectType": "Content"
                },
                {
                    "identifier": "obj-cat:textbook-unit_collection_all",
                    "name": "Textbook Unit",
                    "targetObjectType": "Collection"
                }
            ],
            "additionalCategories": [
                "Asset",
                "CCT",
                "CertAsset",
                "Certificate Template",
                "Classroom Teaching Video",
                "Concept Map",
                "Content Playlist",
                "Course",
                "Course Assessment",
                "Course Unit",
                "Curiosity Question Set",
                "Digital Textbook",
                "eTextbook",
                "Experiential Resource",
                "Explanation Content",
                "Explanation Video",
                "FLM",
                "Focus Spot",
                "FTB Question",
                "Goals",
                "Good Practices",
                "Learning Outcome Definition",
                "Learning Resource",
                "Lesson Plan",
                "Lesson Plan Unit",
                "Manuals/SOPs",
                "Marking Scheme Rubric",
                "Multiple Choice Question",
                "Pedagogy Flow",
                "Playlist",
                "Plugin",
                "Practice Question Set",
                "Previous Board Exam Papers",
                "Reports",
                "Subjective Question",
                "Teacher Resource",
                "Template",
                "Textbook",
                "Textbook Unit",
                "TV Lesson"
            ],
            "lastUpdatedOn": "2024-05-29T11:08:08.820+0000",
            "collectionAdditionalCategories": [
                "Textbook",
                "Lesson Plan"
            ],
            "assetPrimaryCategories": [
                "Asset",
                "CertAsset",
                "Certificate Template"
            ],
            "contentAdditionalCategories": [
                "Classroom Teaching Video",
                "Concept Map",
                "Curiosity Question Set",
                "Experiential Resource",
                "Explanation Video",
                "Focus Spot",
                "Learning Outcome Definition",
                "Lesson Plan",
                "Marking Scheme Rubric",
                "Pedagogy Flow",
                "Previous Board Exam Papers",
                "TV Lesson",
                "Textbook"
            ],
            "status": "Live",
            "defaultFramework": "NCF"
        }
    }
}`

**Reply by amitpriyadarshi** [reply]
@nidhikachhawa24 it seems the primary category you are passing in the list (i.e. `"Course Assessment", "eTextbook", "Explanation Content", "Learning Resource", "Practice Question Set", "Teacher Resource", "Exam Question", "Lesson Plan Unit"`), not synced into Elastic Search while creation of the primary category. Please validate by calling search api whether all are synced into Elastic Search or not. If not, please sync all of them (also cross check with devops why objects are not getting synced into ES).
cc: @ashokreddy1208

**Comment by anisha147** [comment]
Hi @paritshivani 

We hope the steps shared by @ashokreddy1208 helped resolve the queries. If yes, could you please close this thread from your end?

cc: @ashokreddy1208

**Comment by HarishGangula** [comment]
As we haven't received a response from you, we will proceed to close this thread. If you have any further questions or require assistance in the future, please don't hesitate to start a new thread. Thank you.
